extensions/amp-story/0.1/media-pool.js

```diff
@@ -457,41 +570,65 @@ export class MediaPool {
    * @private
    */
   swapPoolMediaElementIntoDom_(domMediaEl, poolMediaEl, sources) {
-    const ampMediaForPoolEl = ampMediaElementFor(poolMediaEl);
-    const ampMediaForDomEl = ampMediaElementFor(domMediaEl);
+    const swapIntoDom = new SwapIntoDomTask(domMediaEl, this.vsync_);
+
     poolMediaEl[REPLACED_MEDIA_PROPERTY_NAME] = domMediaEl.id;
 
-    return this.enqueueMediaElementTask_(poolMediaEl,
-        new SwapIntoDomTask(domMediaEl, this.vsync_))
-        .then(() => {
-          this.maybeResetAmpMedia_(ampMediaForPoolEl);
-          this.maybeResetAmpMedia_(ampMediaForDomEl);
-          this.enqueueMediaElementTask_(poolMediaEl,
-              new UpdateSourcesTask(sources, this.vsync_));
-          this.enqueueMediaElementTask_(poolMediaEl, new LoadTask());
-        }, () => {
-          this.forceDeallocateMediaElement_(poolMediaEl);
-        });
-  }
+    return this.enqueueMediaElementTask_(poolMediaEl, swapIntoDom).then(() => {
+      const updateSources = new UpdateSourcesTask(sources, this.vsync_);
+      const load = new LoadTask();
+
+      this.enqueueMediaElementTask_(poolMediaEl, updateSources).then(() => {
```

Comment:
`then`ing a promise within a Promise is an antipattern. Return the inner promise, then handle the result in a `then` on the outer Promise chain.

Reply:
I agree that it's not ideal, but in this case moving the callback would have different semantics, given the `catch` callback.